File: packages/guards/src/model-fingerprint.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ModelFingerprintOptions {
  action: 'block' | 'warn';
}

const FINGERPRINT_PATTERNS = [
  /\b(?:I am|I'm)\s+(?:an?\s+)?(?:AI|language model|LLM|GPT|Claude|ChatGPT|Gemini|Llama)\b/i,
  /\b(?:trained by|developed by|created by)\s+(?:OpenAI|Anthropic|Google|Meta|Microsoft)\b/i,
  /\b(?:GPT-[34]|Claude\s+\d|Gemini\s+\d|Llama\s+\d)\b/i,
  /\b(?:my training data|my training cutoff|my knowledge cutoff)\b/i,
  /\b(?:as a large language model|as an AI assistant)\b/i,
];

export function modelFingerprint(options: ModelFingerprintOptions): Guard {
  return {
    name: 'model-fingerprint',
    version: '0.1.0',
    description: 'Detects model identity/fingerprint leakage in output',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];
      for (const p of FINGERPRINT_PATTERNS) { const m = text.match(p); if (m) matched.push(m[0]); }
      const triggered = matched.length > 0;
      return { guardName: 'model-fingerprint', passed: !triggered, action: triggered ? options.action : 'allow', latencyMs: Math.round(performance.now() - start), details: triggered ? { matched } : undefined };
    },
  };
}

File: python/open_guardrail/guards/model_fingerprint.py
"""Detects AI model fingerprint phrases."""

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_PATTERNS = [
    re.compile(r"\bI\s+am\s+an\s+AI\b", re.I),
    re.compile(r"\btrained\s+by\s+OpenAI\b", re.I),
    re.compile(r"\bGPT-4\b"),
    re.compile(r"\bClaude\b"),
    re.compile(r"\bas\s+a\s+large\s+language\s+model\b", re.I),
    re.compile(r"\bmy\s+training\s+data\b", re.I),
]


class _ModelFingerprint:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "model-fingerprint"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: List[str] = []
        for p in _PATTERNS:
            m = p.search(text)
            if m:
                matched.append(m.group()[:60])
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name=self.name,
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="Model fingerprint detected" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched": matched} if triggered else None,
        )


def model_fingerprint(*, action: str = "warn") -> _ModelFingerprint:
    return _ModelFingerprint(action=action)
